### PR TITLE
Ctr to git

### PR DESCRIPTION
**Pull request by MarieLataretu**
- control sequences included in repository
- '--own' can now be a list of files
- '--own', '--host' and '--control' can be combined in whatever way

solves #6 

(only tested with conda)

### CODE REVIEW

File: clean.nf [hunk 1]
```diff
@@ -39,12 +39,14 @@ if( !nextflow.version.matches('20.01+') ) {
     exit 1
 }
 
-if (params.profile) { exit 1, "--profile is WRONG use -profile" }
-if (params.nano == '' &&  params.illumina == '' && params.fasta == '' ) { exit 1, "input missing, use [--nano] or [--illumina] or [--fasta]"}
-if (params.control) {if (params.control != 'phix' &&  params.control != 'dcs' &&  params.control != 'eno') { exit 1, "wrong control defined, use [phix], [dcs] or [eno]"}}
-if (!params.host && !params.own && !params.control) { exit 1, "please provide a control (--control), a host tag (--host) or a FASTA file (--own) for the clean up. A control can be combined with either --host or --own."}
-if (params.host && params.own) {print "Attention, you provided a host via the --host flag (${params.host})\n and your own reference sequence via --own (${params.own}). Only your own file will be used.\n If you provided a control via --control (${params.control}) this will be added to your --own sequence.\n\n"}
-if (params.bbduk_kmer && !params.bbduk) { print "parameter --bbduk_kmer is ignored. Use --bbduk to switch from minimap2 to bbduk for illumina reads.\n\n" }
+Set controls = ['phix', 'dcs', 'eno']
```
Comment (hoelzer) [inline clean.nf:42]: ahh, that's a cool way to check this be defining ``Set``s!

**Comment by hoelzer** [comment]
tested:

```bash
nextflow run clean.nf --nano test/nanopore.fastq.gz --host eco --control dcs
```

running into 

```
No such variable: checkedOwn

 -- Check script 'clean.nf' at line: 140 or see '.nextflow.log' file for more details
```

solved via adding an empty channel

```
      if (params.own) {
        check_own(ownFastaChannel)
        checkedOwn = check_own.out
      }
      else {
        checkedOwn = Channel.empty()
      }
```

**Comment by hoelzer** [comment]
I also run the above command with no specified ``-profile`` so I think the docker was used and working

**Comment by hoelzer** [comment]
At the moment I can only define one control, right? 

```bash
Wrong control defined, use one of these: [phix, dcs, eno]
(base) martin@pop-os:~/git/clean$ nextflow run clean.nf --illumina 'test/illumina*.R{1,2}.fastq.gz' --nano test/nanopore.fastq.gz --fasta test/assembly.fasta --host mmu --control phix,dcs
```

When I give the script Illumina and ONT data I likely want to clean the Illumina data for phiX and the ONT data for DCS. The simplest way would be to concatenate all into the reference index. The more clean way would be to have one index for ONT (including the host, ... and DCS) and one index for Illumina (including the host, ... and phiX). 

@MarieLataretu For me it's also fine for now to go with the more simple way and concatenate an index with host,phiX,DCS except you see a straigth-forward way to build different indices in such a case?

**Comment by hoelzer** [comment]
Also, it seems that for such a command:

```bash
nextflow run clean.nf --illumina 'test/illumina*.R{1,2}.fastq.gz' --nano test/nanopore.fastq.gz --fasta test/assembly.fasta --host mmu --control phix
```

the minimap2 procedure for Illumina is used:

```bash
[71/f88245] process > prepare_host:download_host        [  0%] 0 of 1
[-        ] process > prepare_host:concat_contamination -
[-        ] process > clean_fasta:minimap2_fasta        -
[-        ] process > clean_nano:minimap2_nano          -
[-        ] process > clean_illumina:minimap2_illumina  -
```

but we want to use ``bbduk`` for illumina always.

**Comment by hoelzer** [comment]
> Also, it seems that for such a command:
> 
> ```shell
> nextflow run clean.nf --illumina 'test/illumina*.R{1,2}.fastq.gz' --nano test/nanopore.fastq.gz --fasta test/assembly.fasta --host mmu --control phix
> ```
> 
> the minimap2 procedure for Illumina is used:
> 
> ```shell
> [71/f88245] process > prepare_host:download_host        [  0%] 0 of 1
> [-        ] process > prepare_host:concat_contamination -
> [-        ] process > clean_fasta:minimap2_fasta        -
> [-        ] process > clean_nano:minimap2_nano          -
> [-        ] process > clean_illumina:minimap2_illumina  -
> ```
> 
> but we want to use `bbduk` for illumina always.

okay forget this comment, I saw that I can activate bbduk via ``--bbduk``

**Comment by hoelzer** [comment]
Wow, bbduk needs a lot of RAM for mammalian genomes :) 

```bash
nextflow run clean.nf --illumina 'test/illumina*.R{1,2}.fastq.gz' --host mmu --control phix --bbduk --memory 48.GB
```

so good we also have the minimap2 option for short-reads.

**Comment by MarieLataretu** [comment]
One can now use a comma separated list for `--control`
Is this also a handy option for `--host`?

To create separate indices with the corresponding control is definitely possible. I want just to clarify the behavior:
- if `illumina` and `phix` in `control` (and `own`) -> host+phix(+own)
- if `nano` and `eno` in `control` (and `own`) -> host+eno(+own)
- if `nano` and `dcs` in `control` (and `own`) -> host+dcs(+own)

What if
- if `fasta` -> all from host, own, control?
- if `nano` and  `eno` in `control` and `dcs` in `control` -> not allowed or host+eno+dcs(+own)?

**Comment by hoelzer** [comment]
> One can now use a comma separated list for `--control`
> Is this also a handy option for `--host`?

Great!

Yeah, I think this is also a nice option for `--host`, Example: let's decontaminate against a set of bacterial genomes. 

> * if `illumina` and `phix` in `control` (and `own`) -> host+phix(+own)
> * if `nano` and `eno` in `control` (and `own`) -> host+eno(+own)
> * if `nano` and `dcs` in `control` (and `own`) -> host+dcs(+own)

Correct.

> What if
> 
> * if `fasta` -> all from host, own, control?

Yes. The idea is that we might miss contaminants when we are only looking at the short- or long-read level. Let's say reads are combined during the assembly and a contig is produced that then can be matched against some contaminant (such as phiX). 

> * if `nano` and  `eno` in `control` and `dcs` in `control` -> not allowed or host+eno+dcs(+own)?

Good point! I would not allow this because `eno` comes from ONT dRNA-Seq and `dcs` comes from ONT DNA-Seq. 

So, in summary, the user might provide Illumina data, ONT data, and a hybrid assembly. And now he wants to decontaminate against a `host` and `phix,dcs`. Then the most meaningful behavior would be to 
* clean the Illumina reads against host+phix
* clean the ONT reads against host+dcs
* clean the assembly FASTA against host+phix+dcs

Do you think that's possible to implement with the current methodology?

**Comment by MarieLataretu** [comment]
Yes, sure, I'm on it!

Should we also exit when `illumina` (and no `nano`) and `dcs` or `eno` in `control`? (and vice versa for `nano` and `phix`)?

**Comment by hoelzer** [comment]
> Yes, sure, I'm on it!
> 
> Should we also exit when `illumina` (and no `nano`) and `dcs` or `eno` in `control`? (and vice versa for `nano` and `phix`)?
Hm no, I think that is fine. When really someone wants to check ONT reads for phiX - why not. :) 

Do we miss something or can I go again through the code and test multiple hosts and then merge the PR?

**Comment by MarieLataretu** [comment]
Okay, now everything should be implemented!

**Comment by hoelzer** [comment]
> Okay, now everything should be implemented!

Great! I successfully run 

```bash
nextflow run clean.nf --nano test/nanopore.fastq.gz --illumina 'test/illumina*.R[1,2].fastq.gz' --fasta test/assembly.fasta --host cli,eco --control phix,dcs --bbduk --cores 6 --memory 52.GB -with-dag figures/dag.png
```

and also generated and added the new flow chart.
